### ISSUE TITLE
trigger=True on dropdown locks up the app

### ISSUE THREAD

**Issue by fccoelho** [h2oai/wave#1684]
#### Wave SDK Version, OS

h20wave 0.23.1 on Ubuntu 22.04, Python 3.10

#### Actual behavior

I have a simple app with a dropdown widget, that I  need to get the value from. But if I set `trigger=True` the app goes into an endless `loading...` screen.

#### Expected behavior
I expected the program not to lock up.

#### Steps To Reproduce

Here is my ui:
```python
def add_sidebar(q):
    state_choices = [
        ui.choice('SC', 'Santa Catarina'),
        ui.choice('RS', 'Rio Grande do Sul')
    ]
    q.page['form'] = ui.form_card(box='sidebar', items=[
        ui.dropdown(name='state', label='Select state', value='RS', required=True,
                    choices=state_choices, trigger=False),
        ui.dropdown(name='city', label='Select city', required=True,
                    choices=[], trigger=True, visible=False)
    ])
```
In the above code I have trigger only on the second dropdown, if I select from the first, the app does not lock, but I don't get the updated value in `q.args.state`. If I select from the second one, the app locks.

**Comment by mturoci** [comment]
Hi @fccoelho and thanks for your interest in Wave.

Actually, this is a feature, not a bug :) Wave expects every user interaction to result in a UI update - a good UX practice. The reason you are getting the infinite loading is that the UI update never came after the dropdown value submission.

We highly recommend updating the UI in response to user interaction to let your users know something is happening. However, if you do not want to do that, there is a hack - simply update non-existent card, e.g. `q.page['non-existent'].items = []` - this should make the loading go away.

I will make this part of our docs as this question is very common.

**Comment by mturoci** [comment]
Fixed in https://github.com/h2oai/wave/commit/c86ec398bbffe1ef7da983550814a5030b01ddad

**Comment by fccoelho** [comment]
> We highly recommend updating the UI in response to user interaction to let your users know something is happening.

Do you mean a `q.page.save()` call? 
I have managed to make it go away after adding an extra `q.page.save()` call to the functions responding to these interactions.

thanks a lot!

**Comment by mturoci** [comment]
> Do you mean a q.page.save() call?

Not necessarily. `q.page.save()` sends all the changes down to server, but if there are no changes, it will not help.

Let's see 2 code examples to make this more clear.

```py
@app('/')
async def serve(q: Q):
    # Run this code only once per browser tab initialization.
    if not q.client.initialized:
        q.page['form'] = ui.form_card(box='1 1 3 3', items=[
            ui.button(name='btn', label='Btn'),		
        ])
        q.client.initialized = True

    print('calling save, but still getting loading')
    await q.page.save()

```

Despite this code calls `q.page.save`, it doesn't help since there are no UI changes.

```py
@app('/')
async def serve(q: Q):
    # Run this code only once per browser tab initialization.
    if not q.client.initialized:
        q.page['form'] = ui.form_card(box='1 1 3 3', items=[
            ui.button(name='btn', label='Btn'),		
        ])
        q.client.initialized = True

    # Handle btn click.
    if q.args.btn:
        # Update UI - add a new card.
        q.page['new_card'] = ui.form_card(box='1 4 3 3', items=[
            ui.text(content='New card as updated UI for example'),		
        ])

    await q.page.save()
```
However, the second code adds a new card every time a button is clicked thus loading is not shown.

Hope this makes it more clear!

**Comment by fccoelho** [comment]
Thanks, I get it now! 

My problems are solved, but I still don't see what's the use case for `trigger=False`

**Comment by mturoci** [comment]
> I still don't see what's the use case for trigger=False

Imagine you have a form with 2 dropdowns. If both have `trigger=True`. If you change dropdown1, its value gets submitted to the Wave server and is ready for you to consume. The same is true for the dropdown2. The problem is when both values must be submitted together in one pass. That's where `trigger=False` and an extra submission button comes into play as clicking the button will submit every "newly changed values on the page", thus both of our dropdowns.

Feel free to direct any more questions to our [Discord](https://discord.com/invite/V8GZFAy3WM) server so that all this knowledge is not buried within the GH issue, but can be helpful to others as well.

**Comment by mturoci** [comment]
Closing as answered.
